docs/proposals/nodesets.md

```diff
@@ -0,0 +1,157 @@
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# NodeSets in Kubernetes
+
+**Author**: Justin Santa Barbara
+
+**Status**: Proposal.
+
+This document presents the design of the Kubernetes NodeSet, describes use cases, and gives an overview of the code.
+
+## Motivation
+
+The official bring-up procedure for a kubernetes cluster is to use kube-up.  This is a pretty complicated bash script
+that is largely separate for each provider.  It offers limited functionality: you cannot change the number of nodes
+directly, nor change the instance type used.  Many people resort to alternative methods of installation,
+to varying degrees of success.
+
+NodeSets put node launching under the control of the master.  They allow the operator to create
+and destroy NodeSets dynamically, or to change the number of instances in a NodeSet.  They
+allow for easier upgrades by launching nodes with different OS/k8s versions.
+
+This simplifies the Kubernetes installation, because the primary requirement is simply to launch a master.  Then
+the installation script can either automatically create a NodeSet, or leave this up to the operator post-creation.
+
+## Use Cases
+
+### Resize number of Nodes through the API
+
+By updating the count on a NodeSet, nodes can be launched or destroyed.  Users will no longer have
+to go to the underlying cloud API.  Kubernetes can likely make better decisions about which instance to
+terminate than a cloud-implemented auto-scaling group can, because it is more aware of the workloads.
+
+### Consistent auto-scaling across providers
+
+In future we can add auto-scaling functionality at the Kubernetes level (instead of at the cloud level), so
+that it works consistently across clouds, and can be based on Kubernetes metrics instead of cloud-metrics
+(i.e. it can take account of pods _not_ scheduled).
+
+### Support multiple NodeSets of different instance types
+
+Users may want to run instances of different types in the same cluster, permanently or transiently.
+Permanently mixing instance types may require more scheduler support, but for transient usage (changing
+instance sizes) this is helpful for changing a cluster.
+
+Users may also want to mix instance types in a way that is not visible to the cluster, for example mixing
+on-demand and spot instances on AWS.
+
+### Support upgrading by specifying software versions
+
+By specifying that a new NodeSet should be launched with a different image or kubernetes version,
+users can easily perform an online upgrade of their nodes (within compatible versions).
+
+## Functionality
+
+The NodeSet supports standard API features:
+- create
+  - The spec for NodeSet includes a count of nodes
+  - The spec includes a template, which specifies the type of nodes that will be launched as labels.
```

Comment:
My experience with volumes has scarred me a bit here: having separate elements for each cloud provider has complicated using volumes.  So my opinion is that I'd rather try to make it work with one set of specifiers; even though some specifier values will only be meaningful on specific clouds.  But I was pondering whether we could have cross-cloud specifiers, by specifying the CPU bogomips or network bandwidth or memory required.  Users could still specify an instance type directly, or they could specify their preferences.  I also debated including the existing resourceSpecification element.

My guess is that for V1 this will be created primarily by kube-up (and equivalents), and most users will just specify an instance type directly.  So my theory was to do the minimal thing for V1 and then allow it to be extended in future.

This is only my opinion/thought though.

My biggest uncertainty is over the use of labels vs typed fields for the specifier.  I think labels is cute, and would be interesting to try: this is going to be in the experimental section at first, so it isn't the end of the world if we decide labels don't work out.